File: rust/src/notification_publisher.rs
//! Notification publisher for sending real-time events to the console server via Unix domain socket.

use crate::error::KanbusError;
use crate::notification_events::NotificationEvent;
use sha2::{Digest, Sha256};
#[cfg(unix)]
use std::io::Write;
#[cfg(unix)]
use std::os::unix::net::UnixStream;
use std::path::{Path, PathBuf};

/// Get the Unix domain socket path for the current project.
///
/// The socket path is derived from the project root directory to ensure
/// each project has its own isolated notification channel.
fn get_socket_path(root: &Path) -> PathBuf {
    let canonical = root.canonicalize().unwrap_or_else(|_| root.to_path_buf());
    let mut hasher = Sha256::new();
    hasher.update(canonical.to_string_lossy().as_bytes());
    let hash = format!("{:x}", hasher.finalize());
    let socket_name = format!("kanbus-{}.sock", &hash[..12]);

    std::env::temp_dir().join(socket_name)
}

/// Publish a notification event to the console server via Unix domain socket.
///
/// This function sends the event to the console server's Unix socket.
/// The socket path is derived from the project root directory to ensure
/// each project has its own isolated notification channel.
///
/// Errors are logged but not propagated - notification failures should
/// not block CRUD operations.
pub fn publish_notification(root: &Path, event: NotificationEvent) -> Result<(), KanbusError> {
    let socket_path = get_socket_path(root);
    let result = send_notification_sync(&socket_path, &event);

    if let Err(e) = result {
        // Log error but don't fail - notification is best-effort
        eprintln!("Warning: Failed to send notification: {}", e);
    }

    Ok(())
}

/// Synchronously send notification via Unix domain socket.
#[cfg(unix)]
fn send_notification_sync(
    socket_path: &Path,
    event: &NotificationEvent,
) -> Result<(), KanbusError> {
    // Try to connect to the Unix socket
    let mut stream = UnixStream::connect(socket_path).map_err(|e| {
        KanbusError::IssueOperation(format!(
            "Console server not reachable (socket: {}): {}",
            socket_path.display(),
            e
        ))
    })?;

    // Serialize event to JSON and send as newline-delimited message
    let json_body = serde_json::to_string(event)
        .map_err(|e| KanbusError::IssueOperation(format!("Failed to serialize event: {}", e)))?;

    stream
        .write_all(json_body.as_bytes())
        .map_err(|e| KanbusError::IssueOperation(format!("Failed to write to socket: {}", e)))?;

    stream
        .write_all(b"\n")
        .map_err(|e| KanbusError::IssueOperation(format!("Failed to write newline: {}", e)))?;

    Ok(())
}

#[cfg(not(unix))]
fn send_notification_sync(
    _socket_path: &Path,
    _event: &NotificationEvent,
) -> Result<(), KanbusError> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_publish_notification_does_not_block() {
        // This should complete immediately even if server is unreachable
        let temp_dir = tempdir().unwrap();
        let event = NotificationEvent::IssueFocused {
            issue_id: "test-123".to_string(),
            user: None,
        };

        let result = publish_notification(temp_dir.path(), event);
        assert!(result.is_ok());
    }

    #[test]
    fn test_socket_path_is_deterministic() {
        let temp_dir = tempdir().unwrap();
        let path1 = get_socket_path(temp_dir.path());
        let path2 = get_socket_path(temp_dir.path());
        assert_eq!(path1, path2);
    }

    #[test]
    fn test_socket_path_differs_by_project() {
        let temp_dir1 = tempdir().unwrap();
        let temp_dir2 = tempdir().unwrap();
        let path1 = get_socket_path(temp_dir1.path());
        let path2 = get_socket_path(temp_dir2.path());
        assert_ne!(path1, path2);
    }
}

File: python/src/kanbus/notification_publisher.py
"""Notification publisher: sends real-time events to the console server via Unix socket."""

from __future__ import annotations

import hashlib
import json
import socket
import tempfile
from pathlib import Path


def _get_socket_path(root: Path) -> Path:
    """Return the Unix domain socket path for the given project root.

    Uses the same derivation as the Rust notification_publisher:
    SHA-256 of the canonical path, first 12 hex characters.

    :param root: Repository root path.
    :type root: Path
    :return: Path to the Unix domain socket.
    :rtype: Path
    """
    try:
        canonical = str(root.resolve())
    except OSError:
        canonical = str(root)
    digest = hashlib.sha256(canonical.encode()).hexdigest()
    socket_name = f"kanbus-{digest[:12]}.sock"
    return Path(tempfile.gettempdir()) / socket_name


def publish_notification(root: Path, event: dict) -> None:  # type: ignore[type-arg]
    """Send a notification event to the running console server.

    Errors are silenced — notification failure must not block CLI operations.

    :param root: Repository root path.
    :type root: Path
    :param event: Notification event dict (will be JSON-serialised).
    :type event: dict
    """
    socket_path = _get_socket_path(root)
    try:
        payload = json.dumps(event) + "\n"
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.settimeout(2.0)
            sock.connect(str(socket_path))
            sock.sendall(payload.encode())
    except (OSError, socket.error):
        pass
